content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md

```diff
@@ -0,0 +1,547 @@
+
+To set up RBAC, you create a Service Account, a ClusterRole, and connect the two with a Cluster RoleBinding. 
+
+#### 1. Create a YAML file *sa.yaml* for the service account:
+
+```markdown
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: mongo-account
+```
+
+#### 2. Create a YAML file *clusterrole.yaml* for the cluster roles:
+
+```markdown
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: mongo-role
+rules:
+- apiGroups: [""]
+  resources: ["deployments"]
+  verbs: ["list", "watch"]
+- apiGroups: [""]
+  resources: ["services"]
+  verbs: ["*"]
+- apiGroups: [""]
+  resources: ["pods"]
+  verbs: ["get","list", "watch"]
+```
+
+#### 3. Create a YAML file *clusterrolebinding.yaml* to bind the service account with the cluster access roles:
+
+```markdown
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: mongo_role_binding
+subjects:
+- kind: ServiceAccount
+  name: mongo-account
+roleRef:
+  kind: ClusterRole
+  name: mongo-role
+  apiGroup: rbac.authorization.k8s.io
+```
+
+### Create the StatefuleSet Deployment
+
+To deploy MongoDB as a Replica Set with multiple pods, a Kubernetes StatefulSet deployment will be required. The data persistence setup can be done with a *VolumeClaimTemplate* in the StatefulSet deployment.
+
+It should be noted that the Kubernetes Deployment works fine if you have only one single MongoDB replica being deployed. In case multiple replicas are running, which is required in the production environment, there will be some issues using Kubernetes Deployment. Developers have to handle the concurrent read-write of the same data in the deployment. 
+
+#### 1. Create a YAML file *namespace.yaml* for the namespace:
+
+```markdown
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: mongodb
+```
+
+#### 2. Create a YAML file *statefulset.yaml* for MongoDB StatefulSet deployment:
+
+```markdown
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: mongodb-replica
+spec:
+  serviceName: pce-mongodb
+  replicas: 1 
+  selector:
+    matchLabels:
+      app: mongo
+  template:
+    metadata:
+      labels:
+        app: mongo
+        selector: mongo
+    spec:
+      terminationGracePeriodSeconds: 30
+      serviceAccount: pce-mongo-account
+      containers:
+      - name: mongodb
+        image: docker.io/mongo:4.2
+        command: ["/bin/sh"]
+        args: ["-c", "mongod --replSet=rs0 --bind_ip_all"]
+        resources:
+          limits:
+            cpu: 100m
+            memory: 200Mi
+          requests:
+            cpu: 100m
+            memory: 100Mi
+        ports:
+        - name: mongo-port
+          containerPort: 27017
+        volumeMounts:
+        - name: mongo-data
+          mountPath: /data/db
+  volumeClaimTemplates:
+  - metadata:
+      name: mongo-data
+    spec:
+      accessModes:
+        - ReadWriteOnce
+      resources:
+        requests:
+          storage: 1Gi
+```
+
+### Set Up the Headless Service
```

Comment:
"Set **up** the Headless Service" (from what I can see, Headless Service is a proper noun, hence the initial capitalization). But you still need to change "up"